src/include/api.h

```diff
@@ -10,48 +10,59 @@
 /*
  * Capture cases where a single session handle is used by multiple threads in parallel. The check
  * isn't trivial because some API calls re-enter via public API entry points and the session with ID
- * 0 is the default session in the connection handle which can be used across multiple threads. It
- * is safe to use the reference count without atomic operations because the reference count is only
- * tracking a thread re-entering the API.
+ * 0 is the default session in the connection handle which can be used across multiple threads.
  */
-#define WT_SINGLE_THREAD_CHECK_START(s)                                                   \
-    {                                                                                     \
-        uintmax_t __tmp_api_tid;                                                          \
-        __wt_thread_id(&__tmp_api_tid);                                                   \
-        WT_ASSERT(s, (s)->id == 0 || (s)->api_tid == 0 || (s)->api_tid == __tmp_api_tid); \
-        if ((s)->api_tid == 0) {                                                          \
-            (s)->api_tid = __tmp_api_tid;                                                 \
-            WT_FULL_BARRIER();                                                            \
-        }                                                                                 \
-        ++(s)->api_enter_refcnt;                                                          \
+#define WT_SINGLE_THREAD_CHECK_START(s)                                           \
+    {                                                                             \
+        uintmax_t __tmp_api_tid;                                                  \
+        __wt_thread_id(&__tmp_api_tid);                                           \
+                                                                                  \
+        /*                                                                        \
+         * Don't attempt take the lock if this is the default session             \
+         * (ID == 0), or if we already hold the lock.                             \
+         */                                                                       \
+        if ((s)->id != 0 && (s)->thread_check.owning_thread != __tmp_api_tid) {   \
+            WT_ASSERT((s), __wt_spin_trylock((s), &(s)->thread_check.lock) == 0); \
```

Comment:
This is kinda weird, the implication is that there should only every be 1 thread trying to do this. It's worth having a comment explaining why you'd assert on a spin_trylock. Also should this be an assert_always?

Reply:
>the implication is that there should only every be 1 thread trying to do this
This makes sense to me. `conn->open_session` is called by individual threads, and as long as they don't share it with another thread or go digging in `conn->sessions` we'll have the one user for the lifetime of the session. I'll add a comment clarifying.

`ASSERT_ALWAYS` is something a few of us have discussed in the past. `_ALWAYS` implies it runs in release mode, but because it's wrapped in an `#ifdef HAVE_DIAGNOSTIC` this assert won't run in release. I don't particularly mind either choice but I can see the argument that `_ALWAYS` is misleading here.

Reply:
Comment updated in [f092cbf](https://github.com/wiredtiger/wiredtiger/pull/9810/commits/f092cbfd2792f60030eb1178b512794120562092). I've added that it's unexpected to see different threads access sequentially as well, but happy to be corrected on that.